app/validators/note_presence_on_purchase_validator.rb

```diff
@@ -0,0 +1,22 @@
+class NotePresenceOnPurchaseValidator
+
+  def initialize(order)
+    @order = order
+  end
+
+  def valid?
+    # Use select instead of none? to make sure we loop over everything
+    invalid_orders = @order.order_details.select do |od|
+      if od.product.user_notes_field_mode.required? && od.note.blank?
+        od.errors.add(:note, :blank) # truthy return
+      end
+    end
```

Comment:
I think this is right, but the object boundaries do feel a little odd to me. I think because the inside of the block feels conceptually like an instance method of order detail (feature envy). But I completely understand not wanting to put more code in order detail, so maybe it's a separate validator?